resources/sugarcrm/sugarsoapprotocol.cpp

```diff
@@ -0,0 +1,89 @@
+
+  Copyright (C) 2017 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+  Authors: Jeremy Entressangle <jeremy.entressangle@kdab.com>
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 2 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "sugarsoapprotocol.h"
+#include "sugarjob.h"
+#include "sugarsession.h"
+#include <QNetworkReply>
+#include <KLocalizedString>
+#include <KDebug>
+
+SugarSoapProtocol::SugarSoapProtocol()
+{
+}
+
+int SugarSoapProtocol::login(const QString &user, const QString &password, QString &sessionId, QString &errorMessage)
+{
+    Q_ASSERT(mSession->soap() != nullptr);
+
+    const QByteArray passwordHash = password.toUtf8();
+
+    KDSoapGenerated::TNS__User_auth userAuth;
+    userAuth.setUser_name(user);
+    userAuth.setPassword(QString::fromAscii(passwordHash));
+    userAuth.setVersion(QLatin1String(".01"));
+
+    KDSoapGenerated::TNS__Set_entry_result entry_result = mSession->soap()->login(userAuth, QLatin1String("FatCRM"));
+    if (entry_result.error().number() == "0") {
+        sessionId = entry_result.id();
+        return KJob::NoError;
+    } else {
+        errorMessage = i18nc("@info:status", "Login for user %1 on %2 failed: %3", user, mSession->host(), mSession->soap()->lastError());
+        int faultcode = mSession->soap()->lastErrorCode();
+        if (faultcode == QNetworkReply::UnknownNetworkError ||
+           faultcode == QNetworkReply::HostNotFoundError) {
+            return SugarJob::CouldNotConnectError;
+        } else {
+            return SugarJob::LoginError;
+        }
+    }
+}
+
+void SugarSoapProtocol::logout()
+{
+    if (mSession->sessionId().isEmpty() && mSession->soap() != nullptr) {
+        KDSoapGenerated::TNS__Error_value errorValue = mSession->soap()->logout(mSession->sessionId());
+        if (errorValue.number() != "0") {
+            kDebug() << "logout returned error" << errorValue.number() << errorValue.name() << errorValue.description();
+        }
+        if (!mSession->soap()->lastError().isEmpty()) {
+            kDebug() << "logout had fault" << mSession->soap()->lastError();
+        }
+    }
+    mSession->forgetSession();
+}
+
+void SugarSoapProtocol::getEntriesCount(const ListEntriesScope &scope, const QString &moduleName, const QString &query)
+{
+    mSession->soap()->asyncGet_entries_count(mSession->sessionId(), moduleName, query, scope.deleted());
```

Comment:
Can't be async (it requires the caller to connect to signals, so it doesn't allow a mock implementation...)

Make this sync, just like you did for login.

This means you need return the following data from this method:
- the entries count (for when it works)
- a KJob error code (for when it fails)
- an error message (for when it fails)

For consistency with login(), you could make the return value the error code, and add int &entriesCount to return the count as an output parameter. This way around (rather than returning the count and having the errorCode as output parameter) is probably better, in case some other operations need to return multiple things (we'll just use multiple output parameters then).